### DISCUSSION TITLE
What would barebone TS example look like?

### DISCUSSION THREAD

**Discussion by Overtorment** [lightningdevkit/rust-lightning#1569] [Help]
subj

**Comment by moneyball** [comment]
Barebones example:
1. Show user wallet balance
2. Show user payment history
3. Allow user to receive bitcoin by generating an invoice (either bolt11 or BIP 21 on-chain address + bolt11)
4. Allow user to send bitcoin by paying an invoice (perhaps v1 MVP just pasting an invoice but v2 would ideally be 1-click payment on a webpage)

**Reply by TheBlueMatt** [reply]
I guess I don't quite understand the question but a "barebones example" doesn't seem like something that would do a ton of work to track payment history and basically be a full wallet? More like just demo how to hook LDK bits up to existing interfaces and the like.

**Reply by moneyball** [reply]
Pretty sure you are the one that added "show payment history" to the simple API we'd discussed :)

**Reply by TheBlueMatt** [reply]
Right, but the question was "how do I use LDK from TypeScript", not "how do I build a full lightning wallet implementation that is self-contained and doesn't rely on external wallet logic from TypeScript" :) https://github.com/orgs/lightningdevkit/discussions/1569#discussioncomment-3022067

**Comment by Overtorment** [comment]
the answer I was looking for is actually:


```
import * as rawldk from "lightningdevkit/bindings.mjs";
import * as ldk from "lightningdevkit";
import * as fs from "fs";

const test = async () => {
    const result = rawldk.CResult_boolLightningErrorZ_ok(true);
    console.assert(rawldk.CResult_boolLightningErrorZ_is_ok(result));
    console.assert(rawldk.CResult_boolLightningErrorZ_get_ok(result));
    rawldk.CResult_boolLightningErrorZ_free(result);
    const second_res = rawldk.CResult_boolLightningErrorZ_ok(false);
    console.assert(rawldk.CResult_boolLightningErrorZ_is_ok(second_res));
    console.assert(!rawldk.CResult_boolLightningErrorZ_get_ok(second_res));
    rawldk.CResult_boolLightningErrorZ_free(second_res);

    return true;
};


async function run_tests(check_leaks) {
    const result = await test();
    console.log("all tests passed: " + result);
    if (result !== true || !check_leaks) { return result; }

    const allocs_finished = new Promise((resolve, reject) => {
        var loop_count = 0;
        const interval_id = setInterval(() => {
            const alloc_count = rawldk.getRemainingAllocationCount();
            if (loop_count % 20 == 0)
                console.log("Remaining LDK allocation count: " + alloc_count);

            // chromium with --js-flags="--expose-gc" exposes a `window.gc()` which we call if we can
            // @ts-ignore window.gc is considered a type error in TS
            if (typeof window !== "undefined" && typeof window.gc !== "undefined") window.gc();

            // Note that there are currently 9 leaks in the above tests. At least some are known - look for XXX in bindings.c
            if (alloc_count <= 10) { clearInterval(interval_id); rawldk.debugPrintRemainingAllocs(); resolve(true); }
            loop_count += 1;
            if (loop_count > 30*2) { clearInterval(interval_id); rawldk.debugPrintRemainingAllocs(); resolve(false); }
        }, 500);
    });
    return allocs_finished;
}


;(async () => {
    const bin = fs.readFileSync('./node_modules/lightningdevkit/liblightningjs.wasm');

    await ldk.initializeWasmFromBinary(bin);
    return await run_tests(true);
})();


```

**Reply by Overtorment** [reply]
this is not a ts example tho, its barebone nodejs, and in this setup my IDE doesn't recognise any ts bindings, and ts compiler refuses to compile.
its allegedly because of incorrect ldk export to npm

**Comment by gr0kchain** [comment]
```
const fs = require("fs");
async function test  () {
  const { run_tests_node } = await import("lightningdevkit/test/tests.mjs");
  var bin = fs.readFileSync("./node_modules/lightningdevkit/liblightningjs.wasm");
  run_tests_node(bin);
}

test();
```

**Reply by Overtorment** [reply]
yes it runs the tests, but how do I import actual LDK internals in TS, in a way that I get typehints, autocomplete etc etc from IDE..?

**Reply by gr0kchain** [reply]
I use vi so not to familiar with that requirement for our purposes.

**Reply by gr0kchain** [reply]
This appears to work in VS Code though? Not sure if this is what you're referring to?

https://pasteboard.co/gck8xjqDueOb.png
